mtags/src/main/scala-3/scala/meta/internal/pc/MetalsInteractive.scala

```diff
@@ -89,4 +94,59 @@ object MetalsInteractive:
     case _ :: rest =>
       contextOfStat(rest, stat, exprOwner, ctx)
 
+  /**
+   * Check if the given `sourcePos` is on the name of enclosing tree.
+   * ```
+   * // For example, if the postion is on `foo`, returns true
+   * def foo(x: Int) = { ... }
+   *      ^
+   *
+   * // On the other hand, it points to non-name position, return false.
+   * def foo(x: Int) = { ... }
+   *  ^
+   * ```
+   * @param path - path to the position given by `Interactive.pathTo`
+   */
+  def isOnName(
+      path: List[Tree],
+      sourcePos: SourcePosition,
+      source: SourceFile
+  )(using Context): Boolean =
+    def contains(tree: Tree): Boolean = tree match
+      case select: Select =>
+        // using `nameSpan` as SourceTree for Select (especially symbolic-infix e.g. `::` of `1 :: Nil`) miscalculate positions
+        select.nameSpan.contains(sourcePos.span)
+      case tree: NameTree =>
+        SourceTree(tree, source).namePos.contains(sourcePos)
+      // TODO: check the positions for NamedArg and Import
+      case _: NamedArg => true
+      case _: Import => true
+      case app: (Apply | TypeApply) => contains(app.fun)
+      case _ => false
+    end contains
+
+    val enclosing = path
+      .dropWhile(t => !t.symbol.exists && !t.isInstanceOf[NamedArg])
+      .headOption
+      .getOrElse(EmptyTree)
+    contains(enclosing)
+  end isOnName
+
+  private lazy val isForName: Set[Name] = Set[Name](
+    StdNames.nme.map,
+    StdNames.nme.withFilter,
+    StdNames.nme.flatMap,
+    StdNames.nme.foreach
+  )
+  def isForSynthetic(gtree: Tree)(using Context): Boolean =
```

Comment:
Scala3 version of https://github.com/scalameta/metals/blob/35efaad91b0c61b827df7119d15f1eb46c475553/mtags/src/main/scala-2/scala/meta/internal/pc/HoverProvider.scala#L308-L324